### PR TITLE
LITESDCARD peripheral driver support

### PR DESCRIPTION
**Pull request by rtucker85**
## Summary
This PR adds:

1. A driver for the LiteX LITESDCARD peripheral
2. Changes to the ARTY_A7 development board to demonstrate SDCARD support

## Impact
Isolated to ARTY_A7 development board users with LITEX_SDIO configuration turned on

## Testing
Built and tested on the [Digilent ARTY_A7](https://digilent.com/shop/arty-a7-artix-7-fpga-development-board/) development board with the [Digilent PmodSD](https://digilent.com/shop/pmod-sd-full-sized-sd-card-slot/) adaptor plugged into PMOD JD.

### CODE REVIEW

**Comment by xiaoxiang781216** [comment]
Please fix the warning report here:
https://github.com/apache/incubator-nuttx/runs/5671271324?check_suite_focus=true

File: arch/risc-v/src/litex/litex_cache.c [hunk 1]
```diff
@@ -0,0 +1,76 @@
...
+ *
+ * Returned Value:
+ *   None
+ *
+ ****************************************************************************/
+
+#ifdef CONFIG_ARCH_DCACHE
+void up_invalidate_dcache_all(void)
+{
+  asm volatile(".word(0x500F)\n");
```
Comment (xiaoxiang781216) [inline arch/risc-v/src/litex/litex_cache.c:46]: why not use the instruction directly

File: arch/risc-v/src/litex/litex_cache.c [hunk 2]
```diff
@@ -0,0 +1,76 @@
...
+ * Returned Value:
+ *   None
+ *
+ ****************************************************************************/
+
+#ifdef CONFIG_ARCH_ICACHE
+void up_invalidate_icache_all(void)
+{
+    asm volatile(
+      ".word(0x100F)\n"
```
Comment (xiaoxiang781216) [inline arch/risc-v/src/litex/litex_cache.c:68]: ditto

File: include/nuttx/sdio.h [hunk 3]
```diff
@@ -477,6 +477,7 @@
 #define SDIO_CAPS_DMABEFOREWRITE  0x04 /* Bit 2=1: Executes DMA before write command */
 #define SDIO_CAPS_4BIT            0x08 /* Bit 3=1: Supports 4 bit operation */
 #define SDIO_CAPS_8BIT            0x10 /* Bit 4=1: Supports 8 bit operation */
+#define SDIO_CAPS_4BIT_ONLY       0x20 /* Bit 5=1: Supports 4-bit only operation */
```
Comment (xiaoxiang781216) [inline include/nuttx/sdio.h:480]: why not remove SDIO_CAPS_8BIT from cap?

Comment (AndrewD) [reply include/nuttx/sdio.h:480]: The issue is that the peripheral does not support 1bit mode, so 4 bit mode must be selected _before_ checking the SCR register.

**Comment by acassis** [comment]
Hi @rtucker85 could you please fix these issues:
```
 ../nuttx/tools/checkpatch.sh -g c92c4af304dcc09254dc4023ca822fef11d98233..HEAD
Error: /home/runner/work/incubator-nuttx/incubator-nuttx/nuttx/arch/risc-v/src/litex/litex_cache.c:2:1: error: Relative file path does not match actual file
Error: /home/runner/work/incubator-nuttx/incubator-nuttx/nuttx/arch/risc-v/src/litex/litex_cache.c:76:1: error: "Private/Public Functions" not found! File will not be checked
Error: /home/runner/work/incubator-nuttx/incubator-nuttx/nuttx/boards/risc-v/litex/arty_a7/src/arty_a7.h:30:1: error: Too many blank lines
Error: /home/runner/work/incubator-nuttx/incubator-nuttx/nuttx/boards/risc-v/litex/arty_a7/src/litex_sdio.c:126:1: error: "Private/Public Functions" not found! File will not be checked
Error: Process completed with exit code 1.
```

File: include/nuttx/sdio.h (hunk 3 above)
Comment (rtucker85) [reply include/nuttx/sdio.h:480]: @AndrewD  Is correct. This peripheral _only_ supports 4 bit mode.
I.e. It doesnt support 1bit _or_ 8bit modes.

We therefore must have a mechanism to force the mmcsd driver to put the card into 4 bit mode without checking the capabilities of the card obtained from the SCR register.

Reading the SCR register becomes redundant but has been left in place for diagnostic purposes - If DEBUG_FS_INFO is enabled.

File: arch/risc-v/src/litex/litex_cache.c (hunk 1 above)
Comment (rtucker85) [reply arch/risc-v/src/litex/litex_cache.c:50]: This is a custom instruction for VexRiscV - It does not appear in the RiscV standard.
Ill make a comment in the code to document this.

File: arch/risc-v/src/litex/litex_cache.c (hunk 2 above)
Comment (rtucker85) [reply arch/risc-v/src/litex/litex_cache.c:72]: ditto

File: arch/risc-v/src/litex/litex_cache.c [hunk 4]
```diff
@@ -0,0 +1,88 @@
+/****************************************************************************
+ * arch/risc-v/src/litex/litex_cache.c
```
Comment (pkarashchenko) [inline arch/risc-v/src/litex/litex_cache.c:2]: Maybe we can create `arch/risc-v/src/litex/litex_cache.S` instead? I mean that there are only inline ASM calls here

File: arch/risc-v/src/litex/litex_sdio.c [hunk 5]
```diff
@@ -0,0 +1,1450 @@
...
+#endif
+
+#ifndef CONFIG_LITEX_SD4BIT_FREQ
+#  define CONFIG_LITEX_SD4BIT_FREQ 50000000  /* 25MHz SD 4-bit, normal clocking */
+#endif
+
+#define max(x, y) (((x) > (y)) ? (x) : (y))
+#define min(x, y) (((x) < (y)) ? (x) : (y))
+
+struct litex_dev_s
```
Comment (pkarashchenko) [inline arch/risc-v/src/litex/litex_sdio.c:83]: Please equip with
```
/****************************************************************************
 * Included Files
 ****************************************************************************/

/****************************************************************************
 * Pre-processor Definitions
 ****************************************************************************/

etc.
```

File: arch/risc-v/src/litex/litex_sdio.c [hunk 6]
```diff
@@ -0,0 +1,1450 @@
...
+  uint8_t            block_size;
+};
+
+static int  litex_interrupt(int irq, void *context, void *arg);
+
+static void litex_reset(FAR struct sdio_dev_s *dev);
+static sdio_capset_t litex_capabilities(FAR struct sdio_dev_s *dev);
+static sdio_statset_t litex_status(FAR struct sdio_dev_s *dev);
+static void litex_widebus(FAR struct sdio_dev_s *dev,
+                bool enable);
```
Comment (pkarashchenko) [inline arch/risc-v/src/litex/litex_sdio.c:124]:
```suggestion
static void litex_widebus(FAR struct sdio_dev_s *dev,
                          bool enable);
```
Here and in other places

File: arch/risc-v/src/litex/litex_sdio.c [hunk 7]
```diff
@@ -0,0 +1,1450 @@
...
+
+  /* check for card change interrupt */
+
+  if (pending & LITEX_INT_CARDDETECT)
+  {
+    mcinfo("Card Detect State: %lu\n", getreg32(LITEX_SDPHY_CARD_DETECT));
+
+    /* Perform callback */
+
+    if (priv->do_sdio_card)
```
Comment (pkarashchenko) [inline arch/risc-v/src/litex/litex_sdio.c:322]:
```suggestion
    if (priv->do_sdio_card != NULL)
```

File: arch/risc-v/src/litex/litex_sdio.c [hunk 8]
```diff
@@ -0,0 +1,1450 @@
...
+  pending = getreg32(LITEX_SDIRQ_PENDING) & getreg32(LITEX_SDIRQ_ENABLE);
+  putreg32(pending, LITEX_SDIRQ_PENDING);
+
+  mcinfo("pending irq: %08" PRIx32 " enabled irq: %08" PRIx32 "\n",
+         getreg32(LITEX_SDIRQ_PENDING),
+         getreg32(LITEX_SDIRQ_ENABLE));
+
+  /* check for card change interrupt */
+
+  if (pending & LITEX_INT_CARDDETECT)
```
Comment (pkarashchenko) [inline arch/risc-v/src/litex/litex_sdio.c:316]:
```suggestion
  if ((pending & LITEX_INT_CARDDETECT) != 0)
```

File: arch/risc-v/src/litex/litex_sdio.c [hunk 9]
```diff
@@ -0,0 +1,1450 @@
...
+
+static void litex_configxfrints(struct litex_dev_s *priv, uint32_t xfrints)
+{
+  irqstate_t flags;
+
+  flags = enter_critical_section();
+  priv->xfrints = xfrints;
+
+  putreg32(priv->xfrints | priv->waitints | priv->cintints,
+    LITEX_SDIRQ_ENABLE);
```
Comment (pkarashchenko) [inline arch/risc-v/src/litex/litex_sdio.c:275]:
```suggestion
  putreg32(priv->xfrints | priv->waitints | priv->cintints,
           LITEX_SDIRQ_ENABLE);
```

File: arch/risc-v/src/litex/litex_sdio.c [hunk 10]
```diff
@@ -0,0 +1,1450 @@
...
+
+        /* Yes.. wake up any waiting threads */
+
+        litex_endwait(priv, SDIOWAIT_TRANSFERDONE);
+      }
+  }
+
+  /* check for DMA read interrupt */
+
+  if (pending & LITEX_INT_MEM2BLOCK)
```
Comment (pkarashchenko) [inline arch/risc-v/src/litex/litex_sdio.c:344]:
```suggestion
  if ((pending & LITEX_INT_MEM2BLOCK) != 0)
```

File: arch/risc-v/src/litex/litex_sdio.c [hunk 11]
```diff
@@ -0,0 +1,1450 @@
...
+  divider = litex_pow2roundup(divider);
+  divider = min(max(divider, 2), 256);
+
+  /* this is the *effective* new clk_freq */
+
+  clk_freq = litex_get_hfclk() / divider;
+  if (clk_freq > 1000000)
+    mcinfo("Setting SDCard clk freq to %ld MHz\n", clk_freq / 1000000);
+  else
+    mcinfo("Setting SDCard clk freq to %ld KHz\n", clk_freq / 1000);
```
Comment (pkarashchenko) [inline arch/risc-v/src/litex/litex_sdio.c:610]: please add `{}`

File: arch/risc-v/src/litex/litex_sdio.c [hunk 12]
```diff
@@ -0,0 +1,1450 @@
...
+
+      case CLOCK_SD_TRANSFER_4BIT:
+        clk_freq = CONFIG_LITEX_SD4BIT_FREQ;
+        break;
+    }
+
+  /* Set the new clock frequency along with the clock enable/disable bit */
+
+  uint32_t divider;
+  divider = clk_freq ? litex_get_hfclk() / clk_freq : 256;
```
Comment (pkarashchenko) [inline arch/risc-v/src/litex/litex_sdio.c:600]: Can we have `#define MAX_DIVIDER 256` and use it in all places?

File: arch/risc-v/src/litex/litex_sdio.c [hunk 13]
```diff
@@ -0,0 +1,1450 @@
...
+  putreg32(arg, LITEX_SDCORE_CMD_ARGUMENT);
+  putreg32(regval, LITEX_SDCORE_CMD_COMMAND);
+  putreg32(1, LITEX_SDCORE_CMD_SEND);
+
+  mcinfo("cmd: %lu cmdid: %lu arg: %08lx regval: %08lx\n",
+    cmd,
+    cmdidx,
+    arg,
+    regval
+  );
```
Comment (pkarashchenko) [inline arch/risc-v/src/litex/litex_sdio.c:732]:
```suggestion
  mcinfo("cmd: %" PRIu32 " cmdid: %" PRIu32 " arg: %08" PRIx32 " regval: %08" PRIx32 "\n",
         cmd, cmdidx, arg, regval);
```

File: arch/risc-v/src/litex/litex_sdio.c [hunk 14]
```diff
@@ -0,0 +1,1450 @@
...
+
+    case MMCSD_WRDATAXFR:
+      transfer = SD_CTL_DATA_XFER_WRITE;
+      break;
+    }
+
+  /* Write the SDIO CMD */
+
+  cmdidx = (cmd & MMCSD_CMDIDX_MASK) >> MMCSD_CMDIDX_SHIFT;
+  regval = cmdidx << 8 | transfer << 5 | resplen;
```
Comment (pkarashchenko) [inline arch/risc-v/src/litex/litex_sdio.c:722]:
```suggestion
  regval = (cmdidx << 8) | (transfer << 5) | resplen;
```

File: arch/risc-v/src/litex/litex_sdio.c [hunk 15]
```diff
@@ -0,0 +1,1450 @@
...
+
+  litex_configxfrints(priv, LITEX_INT_BLOCK2MEM);
+
+  /* flush CPU d-cache */
+
+  up_invalidate_dcache_all();
+
+  putreg32(0, LITEX_SDBLOCK2MEM_DMA_ENABLE);
+  putreg32((uint64_t)(uintptr_t)buffer >> 32, LITEX_SDBLOCK2MEM_DMA_BASE);
+  putreg32((uint64_t)(uintptr_t)buffer, LITEX_SDBLOCK2MEM_DMA_BASE + 0x04);
```
Comment (pkarashchenko) [inline arch/risc-v/src/litex/litex_sdio.c:840]:
```suggestion
  putreg32((uintptr_t)buffer >> 32, LITEX_SDBLOCK2MEM_DMA_BASE);
  putreg32((uintptr_t)buffer, LITEX_SDBLOCK2MEM_DMA_BASE + 0x04);
```

File: arch/risc-v/src/litex/litex_sdio.c [hunk 16]
```diff
@@ -0,0 +1,1450 @@
...
+ *   buffer - The memory to DMA into
+ *   nbytes - The size of the DMA transfer in bytes
+ *
+ * Returned Value:
+ *   OK on success; a negated errno on failure
+ *
+ ****************************************************************************/
+
+static int litex_sendsetup(FAR struct sdio_dev_s *dev,
+                              FAR const uint8_t *buffer, size_t nbytes)
```
Comment (pkarashchenko) [inline arch/risc-v/src/litex/litex_sdio.c:867]:
```suggestion
static int litex_sendsetup(struct sdio_dev_s *dev,
                           const uint8_t *buffer, size_t nbytes)
```

File: arch/risc-v/src/litex/litex_sdio.c [hunk 17]
```diff
@@ -0,0 +1,1450 @@
...
+
+  litex_configxfrints(priv, LITEX_INT_MEM2BLOCK);
+
+  /* flush CPU d-cache */
+
+  up_invalidate_dcache_all();
+
+  putreg32(0, LITEX_SDMEM2BLOCK_DMA_ENABLE);
+  putreg32((uint64_t)(uintptr_t)buffer >> 32, LITEX_SDMEM2BLOCK_DMA_BASE);
+  putreg32((uint64_t)(uintptr_t)buffer, LITEX_SDMEM2BLOCK_DMA_BASE + 0x04);
```
Comment (pkarashchenko) [inline arch/risc-v/src/litex/litex_sdio.c:883]:
```suggestion
  putreg32((uintptr_t)buffer >> 32, LITEX_SDMEM2BLOCK_DMA_BASE);
  putreg32((uintptr_t)buffer, LITEX_SDMEM2BLOCK_DMA_BASE + 0x04);
```

File: arch/risc-v/src/litex/litex_sdio.c [hunk 18]
```diff
@@ -0,0 +1,1450 @@
...
+ * Input Parameters:
+ *   dev  - An instance of the SDIO device interface
+ *   cmd  - The command that was sent.  See 32-bit command definitions above.
+ *
+ * Returned Value:
+ *   OK is success; a negated errno on failure
+ *
+ ****************************************************************************/
+
+static int litex_waitresponse(FAR struct sdio_dev_s *dev, uint32_t cmd)
```
Comment (pkarashchenko) [inline arch/risc-v/src/litex/litex_sdio.c:905]:
```suggestion
static int litex_waitresponse(struct sdio_dev_s *dev, uint32_t cmd)
```

File: arch/risc-v/src/litex/litex_sdio.c [hunk 19]
```diff
@@ -0,0 +1,1450 @@
...
+
+  /* SD_TIMEOUT */
+
+  if (ev & 0x04)
+    return -ETIMEDOUT;
+
+  /* SD_CRCERROR */
+
+  if (ev & 0x08)
+    return -EILSEQ;
```
Comment (pkarashchenko) [inline arch/risc-v/src/litex/litex_sdio.c:937]: Please add `{}`

File: arch/risc-v/src/litex/litex_sdio.c [hunk 20]
```diff
@@ -0,0 +1,1450 @@
...
+
+    if (ev & 0x01)
+      break;
+
+    nxsig_usleep(10);
+  }
+
+  /* SD_WRERROR */
+
+  if (ev & 0x02)
```
Comment (pkarashchenko) [inline arch/risc-v/src/litex/litex_sdio.c:926]: are there any defines for those numbers?

File: arch/risc-v/src/litex/litex_sdio.c [hunk 21]
```diff
@@ -0,0 +1,1450 @@
...
+        ret = -ETIMEDOUT;
+      }
+      else if ((regval & 0x08) != 0)
+      {
+        mcerr("ERROR: CRC failure: %08" PRIx32 "\n", regval);
+        ret = -EIO;
+      }
+    }
+
+  if (rshort)
```
Comment (pkarashchenko) [inline arch/risc-v/src/litex/litex_sdio.c:1041]:
```suggestion
  if (rshort != NULL)
```

File: arch/risc-v/src/litex/litex_sdio.c [hunk 22]
```diff
@@ -0,0 +1,1450 @@
...
+ *   dev      - An instance of the SDIO device interface
+ *   eventset - A bitset of events to enable or disable (see SDIOWAIT_*
+ *              definitions). 0=disable; 1=enable.
+ *
+ * Returned Value:
+ *   None
+ *
+ ****************************************************************************/
+
+static void litex_waitenable(FAR struct sdio_dev_s *dev,
```
Comment (pkarashchenko) [inline arch/risc-v/src/litex/litex_sdio.c:1073]:
```suggestion
static void litex_waitenable(struct sdio_dev_s *dev,
```

File: arch/risc-v/src/litex/litex_sdio.c [hunk 23]
```diff
@@ -0,0 +1,1450 @@
...
+  if ((priv->waitevents & SDIOWAIT_TIMEOUT) != 0)
+    {
+      int delay;
+      int ret;
+
+      /* Yes.. Handle a cornercase: The user requested a timeout event but
+       * with timeout == 0?
+       */
+
+      if (!timeout)
```
Comment (pkarashchenko) [inline arch/risc-v/src/litex/litex_sdio.c:1093]:
```suggestion
      if (timeout == 0)
```

File: arch/risc-v/src/litex/litex_sdio.c [hunk 24]
```diff
@@ -0,0 +1,1450 @@
...
+ *             timeout with no wait.  The timeout value is ignored if
+ *             SDIOWAIT_TIMEOUT is not included in the waited-for eventset.
+ *
+ * Returned Value:
+ *   Event set containing the event(s) that ended the wait.  Should always
+ *   be non-zero.  All events are disabled after the wait concludes.
+ *
+ ****************************************************************************/
+
+static sdio_eventset_t litex_eventwait(FAR struct sdio_dev_s *dev)
```
Comment (pkarashchenko) [inline arch/risc-v/src/litex/litex_sdio.c:1132]:
```suggestion
static sdio_eventset_t litex_eventwait(struct sdio_dev_s *dev)
```

**Review by pkarashchenko — COMMENTED** [review]
Please remove all `FAR`

File: boards/risc-v/litex/arty_a7/src/litex_sdio.c [hunk 25]
```diff
@@ -0,0 +1,134 @@
...
+ *
+ ****************************************************************************/
+
+int litex_sdio_initialize(void)
+{
+  int ret = 0;
+
+  finfo("Initializing SDIO slot %d\n", SDIO_SLOTNO);
+  g_sdio_dev = sdio_initialize(SDIO_SLOTNO);
+  if (!g_sdio_dev)
```
Comment (pkarashchenko) [inline boards/risc-v/litex/arty_a7/src/litex_sdio.c:107]:
```suggestion
  if (g_sdio_dev == NULL)
```

File: boards/risc-v/litex/arty_a7/src/litex_sdio.c [hunk 26]
```diff
@@ -0,0 +1,134 @@
...
+ *
+ ****************************************************************************/
+
+static int litex_sdio_card_isr_callback(void *arg)
+{
+  bool cd;
+
+  cd = litex_cardinserted(SDIO_SLOTNO);
+  finfo("Card detect: %d\n", cd);
+  sdio_mediachange(g_sdio_dev, cd);
```
Comment (pkarashchenko) [inline boards/risc-v/litex/arty_a7/src/litex_sdio.c:65]: why not to get rid of `g_sdio_dev` and pass return of `sdio_initialize` as an `arg` to `litex_sdio_card_isr_callback`?

File: boards/risc-v/litex/arty_a7/src/arty_a7.h [hunk 27]
```diff
@@ -27,6 +27,113 @@
...
+ *    Interrupts are disabled.
+ *
+ ****************************************************************************/
+
+#ifdef HAVE_AUTOMOUNTER
+void litex_automount_event(int slotno, bool inserted);
+#endif
+
+#ifdef HAVE_SDMMC
+bool litex_cardinserted(int slotno);
```
Comment (pkarashchenko) [inline boards/risc-v/litex/arty_a7/src/arty_a7.h:136]: do we need a comment description for this function?

File: arch/risc-v/src/litex/litex_sdio.h [hunk 28]
```diff
@@ -0,0 +1,105 @@
...
+
+void sdio_mediachange(FAR struct sdio_dev_s *dev, bool cardinslot);
+
+#undef EXTERN
+#if defined(__cplusplus)
+}
+#endif
+
+#endif /* __ASSEMBLY__ */
+#endif /* __ARCH_RISCV_SRC_LITEX_LITEX_SDIO_H */
```
Comment (pkarashchenko) [inline arch/risc-v/src/litex/litex_sdio.h:105]:
```suggestion
#endif /* __ARCH_RISCV_SRC_LITEX_LITEX_SDIO_H */

```

File: arch/risc-v/src/litex/litex_sdio.h [hunk 29]
```diff
@@ -0,0 +1,105 @@
...
+#undef EXTERN
+#if defined(__cplusplus)
+#define EXTERN extern "C"
+extern "C"
+{
+#else
+#define EXTERN extern
+#endif
+
+bool litex_sdio_get_card_detect(void);
```
Comment (pkarashchenko) [inline arch/risc-v/src/litex/litex_sdio.h:39]: missing function description

File: arch/risc-v/src/litex/litex_sdio.c [hunk 30]
```diff
@@ -0,0 +1,1450 @@
...
+ ****************************************************************************/
+
+bool litex_sdio_get_card_detect(void)
+{
+  if (!getreg32(LITEX_SDPHY_CARD_DETECT))
+    {
+      return true;
+    }
+
+  return false;
```
Comment (pkarashchenko) [inline arch/risc-v/src/litex/litex_sdio.c:1325]:
```suggestion
  return getreg32(LITEX_SDPHY_CARD_DETECT) == 0;
```

File: arch/risc-v/src/litex/litex_sdio.c (hunk 5 above)
Comment (rtucker85) [reply arch/risc-v/src/litex/litex_sdio.c:98]: Done

File: arch/risc-v/src/litex/litex_sdio.c (hunk 6 above)
Comment (rtucker85) [reply arch/risc-v/src/litex/litex_sdio.c]: Done

File: arch/risc-v/src/litex/litex_sdio.c (hunk 7 above)
Comment (rtucker85) [reply arch/risc-v/src/litex/litex_sdio.c]: Done

File: arch/risc-v/src/litex/litex_sdio.c (hunk 8 above)
Comment (rtucker85) [reply arch/risc-v/src/litex/litex_sdio.c]: Done

File: arch/risc-v/src/litex/litex_sdio.c (hunk 9 above)
Comment (rtucker85) [reply arch/risc-v/src/litex/litex_sdio.c]: Done

File: arch/risc-v/src/litex/litex_sdio.c (hunk 11 above)
Comment (rtucker85) [reply arch/risc-v/src/litex/litex_sdio.c]: Done

File: arch/risc-v/src/litex/litex_sdio.c (hunk 12 above)
Comment (rtucker85) [reply arch/risc-v/src/litex/litex_sdio.c]: Yep! Good idea. Done

File: arch/risc-v/src/litex/litex_sdio.c (hunk 13 above)
Comment (rtucker85) [reply arch/risc-v/src/litex/litex_sdio.c]: Done

File: arch/risc-v/src/litex/litex_sdio.c (hunk 14 above)
Comment (rtucker85) [reply arch/risc-v/src/litex/litex_sdio.c]: Done

File: arch/risc-v/src/litex/litex_sdio.c (hunk 16 above)
Comment (rtucker85) [reply arch/risc-v/src/litex/litex_sdio.c]: Done

File: arch/risc-v/src/litex/litex_sdio.c (hunk 18 above)
Comment (rtucker85) [reply arch/risc-v/src/litex/litex_sdio.c]: Done

File: arch/risc-v/src/litex/litex_sdio.c (hunk 19 above)
Comment (rtucker85) [reply arch/risc-v/src/litex/litex_sdio.c]: Done

File: arch/risc-v/src/litex/litex_sdio.c (hunk 20 above)
Comment (rtucker85) [reply arch/risc-v/src/litex/litex_sdio.c]: Yes there is. Added.

File: arch/risc-v/src/litex/litex_sdio.c (hunk 21 above)
Comment (rtucker85) [reply arch/risc-v/src/litex/litex_sdio.c]: Done

File: arch/risc-v/src/litex/litex_sdio.c (hunk 22 above)
Comment (rtucker85) [reply arch/risc-v/src/litex/litex_sdio.c]: Done

File: arch/risc-v/src/litex/litex_sdio.c (hunk 23 above)
Comment (rtucker85) [reply arch/risc-v/src/litex/litex_sdio.c]: Done

File: arch/risc-v/src/litex/litex_sdio.c (hunk 24 above)
Comment (rtucker85) [reply arch/risc-v/src/litex/litex_sdio.c]: Done

File: boards/risc-v/litex/arty_a7/src/litex_sdio.c (hunk 25 above)
Comment (rtucker85) [reply boards/risc-v/litex/arty_a7/src/litex_sdio.c]: Done

File: boards/risc-v/litex/arty_a7/src/litex_sdio.c (hunk 26 above)
Comment (rtucker85) [reply boards/risc-v/litex/arty_a7/src/litex_sdio.c]: Another good idea. Done.

File: boards/risc-v/litex/arty_a7/src/arty_a7.h (hunk 27 above)
Comment (rtucker85) [reply boards/risc-v/litex/arty_a7/src/arty_a7.h:145]: Added

File: arch/risc-v/src/litex/litex_sdio.h (hunk 28 above)
Comment (rtucker85) [reply arch/risc-v/src/litex/litex_sdio.h]: Done

File: arch/risc-v/src/litex/litex_sdio.h (hunk 29 above)
Comment (rtucker85) [reply arch/risc-v/src/litex/litex_sdio.h:39]: Added

File: arch/risc-v/src/litex/litex_sdio.c (hunk 30 above)
Comment (rtucker85) [reply arch/risc-v/src/litex/litex_sdio.c:1356]: Done

File: arch/risc-v/src/litex/litex_cache.c (hunk 4 above)
Comment (rtucker85) [reply arch/risc-v/src/litex/litex_cache.c]: Ok. Sounds good.
Converted to .S

**Comment by pkarashchenko** [comment]
Thank you for fixing all the comments.
Please rebase your changes on top of the latest mainline to fix build error.

File: arch/risc-v/src/litex/Make.defs [hunk 31]
```diff
@@ -30,7 +30,7 @@ CMN_CSRCS += riscv_initialize.c riscv_swint.c
 CMN_CSRCS += riscv_allocateheap.c riscv_createstack.c riscv_exit.c
 CMN_CSRCS += riscv_assert.c riscv_blocktask.c riscv_copystate.c riscv_initialstate.c
 CMN_CSRCS += riscv_interruptcontext.c riscv_modifyreg32.c riscv_puts.c riscv_mdelay.c
-CMN_CSRCS += riscv_releasepending.c riscv_reprioritizertr.c
+CMN_CSRCS += riscv_releasepending.c riscv_reprioritizertr.c riscv_copyfullstate.c
```
Comment (pkarashchenko) [inline arch/risc-v/src/litex/Make.defs:33]:
```suggestion
CMN_CSRCS += riscv_releasepending.c riscv_reprioritizertr.c
```

**Review by pkarashchenko — COMMENTED** [review]
I see that changes are rebased

File: arch/risc-v/src/litex/litex_cache.S [hunk 32]
```diff
@@ -0,0 +1,82 @@
...
+	.globl  up_invalidate_icache_all
+	.type	  up_invalidate_icache_all, function
+
+up_invalidate_dcaup_invalidate_icache_allhe_all:
+  .word 0x100F
+  nop
+  nop
+  nop
+  nop
+  nop
```
[the remark this answers is not in the document]
Comment (rtucker85) [reply arch/risc-v/src/litex/litex_cache.S]: Sorry about that. That crept in by mistake. All fixed now.

File: arch/risc-v/src/litex/litex_sdio.c [hunk 33]
```diff
@@ -0,0 +1,1487 @@
...
+  struct work_s      cbwork;          /* Callback work queue structure */
+
+  /* Interrupt mode data transfer support */
+
+  uint32_t           xfrints;         /* Interrupt enables for data transfer */
+
+  /* Card interrupt support for SDIO */
+
+  uint32_t           cintints;                /* Interrupt enables for card ints */
+  int               (*do_sdio_card)(void *);  /* SDIO card ISR */
```
[the remark this answers is not in the document]
Comment (rtucker85) [reply arch/risc-v/src/litex/litex_sdio.c]: Good point, we dont. The return value has been removed.

File: arch/risc-v/src/litex/litex_sdio.c (hunk 30 above)
Comment (rtucker85) [reply arch/risc-v/src/litex/litex_sdio.c]: My mistake. This got missed. Actually done now.

File: arch/risc-v/src/litex/litex_sdio.h (hunk 29 above)
Comment (rtucker85) [reply arch/risc-v/src/litex/litex_sdio.h:54]: My mistake. This got missed. Actually added now.

File: arch/risc-v/src/litex/Make.defs (hunk 31 above)
Comment (rtucker85) [reply arch/risc-v/src/litex/Make.defs]: Done

**Review by pkarashchenko — APPROVED** [review]
LGTM! Thank you for contributing this change!

**Review by xiaoxiang781216 — APPROVED** [review]

**Pull request merged**